File: benchmarks/pandas/bench_read_sql_table.py
"""Benchmark: read_sql_table equivalent — read an entire named table into a DataFrame.

Uses an in-memory SQLite database to mirror the tsb readSqlTable benchmark.
`pandas.read_sql_query("SELECT * FROM sensors", con)` is used as the functional
equivalent — both read all rows from a named table via a SQL connection.

Dataset: 10k rows, 3 columns (id int, score float, category str).
Outputs JSON: {"function": "read_sql_table", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import math
import sqlite3
import time

import pandas as pd

ROWS = 10_000
WARMUP = 5
ITERATIONS = 30

# ── Build matching dataset ────────────────────────────────────────────────────
data = {
    "id": list(range(ROWS)),
    "score": [math.sin(i * 0.01) * 100 for i in range(ROWS)],
    "category": [f"cat_{i % 50}" for i in range(ROWS)],
}
df_src = pd.DataFrame(data)

# ── SQLite in-memory database ─────────────────────────────────────────────────
con = sqlite3.connect(":memory:")
df_src.to_sql("sensors", con, index=False, if_exists="replace")

# ── Warm-up ───────────────────────────────────────────────────────────────────
for _ in range(WARMUP):
    pd.read_sql_query("SELECT * FROM sensors", con)

# ── Benchmark ────────────────────────────────────────────────────────────────
start = time.perf_counter()
for _ in range(ITERATIONS):
    pd.read_sql_query("SELECT * FROM sensors", con)
total_ms = (time.perf_counter() - start) * 1000

con.close()

print(json.dumps({
    "function": "read_sql_table",
    "mean_ms": round(total_ms / ITERATIONS, 4),
    "iterations": ITERATIONS,
    "total_ms": round(total_ms, 4),
}))

File: benchmarks/tsb/bench_read_sql_table.ts
/**
 * Benchmark: readSqlTable — read an entire named table into a DataFrame.
 *
 * Uses a mock SqlConnection adapter that returns a 10k-row, 3-column result set.
 * Covers the `readSqlTable` path (table-name validation via `listTables()` +
 * `SELECT * FROM <table>` query dispatch).
 *
 * Outputs JSON: {"function": "read_sql_table", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { readSqlTable } from "../../src/index.js";
import type { SqlConnection, SqlResult, SqlRow, SqlValue } from "../../src/index.js";

const ROWS = 10_000;
const WARMUP = 5;
const ITERATIONS = 30;

// ── Shared result set ─────────────────────────────────────────────────────────
const columns: string[] = ["id", "score", "category"];
const rows: SqlRow[] = Array.from({ length: ROWS }, (_, i) => ({
  id: i,
  score: Math.sin(i * 0.01) * 100,
  category: `cat_${i % 50}`,
}));

// ── Mock adapter ──────────────────────────────────────────────────────────────
class TableAdapter implements SqlConnection {
  query(_sql: string, _params?: readonly SqlValue[]): SqlResult {
    return { columns, rows };
  }
  listTables(): readonly string[] {
    return ["sensors", "events", "logs"];
  }
}

const conn = new TableAdapter();

// ── Warm-up ───────────────────────────────────────────────────────────────────
for (let i = 0; i < WARMUP; i++) {
  readSqlTable("sensors", conn);
}

// ── Benchmark ────────────────────────────────────────────────────────────────
const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  readSqlTable("sensors", conn);
}
const total_ms = performance.now() - start;
const mean_ms = total_ms / ITERATIONS;

console.log(
  JSON.stringify({
    function: "read_sql_table",
    mean_ms: parseFloat(mean_ms.toFixed(4)),
    iterations: ITERATIONS,
    total_ms: parseFloat(total_ms.toFixed(4)),
  }),
);
